homeassistant/components/recorder/history.py

```diff
@@ -59,15 +60,69 @@
     literal(value=None, type_=Text).label("attributes"),
     literal(value=None, type_=Text).label("shared_attrs"),
 ]
+# Remove QUERY_STATES_PRE_SCHEMA_25
+# and the migration_in_progress check
+# once schema 26 is created
+QUERY_STATES_PRE_SCHEMA_25 = [
+    *BASE_STATES,
+    States.attributes,
+    literal(value=None, type_=Text).label("shared_attrs"),
+]
 QUERY_STATES = [
     *BASE_STATES,
+    # Remove States.attributes once all attributes are in StateAttributes.shared_attrs
     States.attributes,
     StateAttributes.shared_attrs,
 ]
 
 HISTORY_BAKERY = "recorder_history_bakery"
 
 
+def query_and_join_attributes(
+    hass: HomeAssistant, no_attributes: bool
+) -> tuple[list[Column], bool]:
+    """Return the query keys and if StateAttributes should be joined."""
+    # If no_attributes was requested we do the query
+    # without the attributes fields and do not join the
+    # state_attributes table
+    if no_attributes:
+        return QUERY_STATE_NO_ATTR, False
+    # If we in the process of migrating schema we do
+    # not want to join the state_attributes table as we
+    # do not know if it will be there yet
+    if recorder.get_instance(hass).migration_in_progress:
+        return QUERY_STATES_PRE_SCHEMA_25, False
+    # Finally if no migration is in progress and no_attributes
+    # was not requested, we query both attributes columns and
+    # join state_attributes
+    return QUERY_STATES, True
+
+
+def bake_query_and_join_attributes(
+    hass: HomeAssistant, no_attributes: bool
+) -> tuple[Any, bool]:
+    """Return the initial backed query and if StateAttributes should be joined.
+
+    Because these are baked queries the values inside the lambdas need
+    to be explicitly written out to avoid caching the wrong values.
+    """
+    bakery: baked.bakery = hass.data[HISTORY_BAKERY]
+    # If no_attributes was requested we do the query
+    # without the attributes fields and do not join the
+    # state_attributes table
+    if no_attributes:
+        return bakery(lambda session: session.query(*QUERY_STATE_NO_ATTR)), False
+    # If we in the process of migrating schema we do
+    # not want to join the state_attributes table as we
+    # do not know if it will be there yet
+    if recorder.get_instance(hass).migration_in_progress:
+        return bakery(lambda session: session.query(*QUERY_STATES_PRE_SCHEMA_25)), False
+    # Finally if no migration is in progress and no_attributes
+    # was not requested, we query both attributes columns and
+    # join state_attributes
+    return bakery(lambda session: session.query(*QUERY_STATES)), True
```

Comment:
I dislike writing this out twice, but I don't think there is another option since the lambda need to be explicitly written out for the bakery